relay-server/src/services/processor.rs

```diff
@@ -1685,29 +1687,53 @@ impl EnvelopeProcessorService {
         }
     }
 
-    /// Returns `true` if the batches should be rate limited.
     #[cfg(feature = "processing")]
-    fn rate_limit_batches(
+    fn rate_limit_buckets_by_namespace(
         &self,
         scoping: Scoping,
-        buckets: &[Bucket],
-        project_state: &ProjectState,
+        buckets: Vec<Bucket>,
+        quotas: &[relay_quotas::Quota],
         mode: ExtractionMode,
-    ) -> bool {
+    ) -> Vec<Bucket> {
         let Some(rate_limiter) = self.inner.rate_limiter.as_ref() else {
-            return false;
+            return buckets;
         };
 
+        let buckets_by_ns: HashMap<MetricNamespace, Vec<Bucket>> = buckets
+            .into_iter()
+            .filter_map(|bucket| Some((bucket.parse_namespace().ok()?, bucket)))
+            .into_group_map();
+
+        buckets_by_ns
+            .into_iter()
+            .filter_map(|(namespace, buckets)| {
+                let item_scoping = ItemScoping {
+                    category: DataCategory::MetricBucket,
+                    scoping: &scoping,
+                    namespace: Some(namespace),
+                };
+
+                (!self.rate_limit_buckets(item_scoping, &buckets, quotas, mode, rate_limiter))
+                    .then_some(buckets)
+            })
+            .flatten()
+            .collect()
```

Comment:
gave it one try and the `BucketsView` taking in a `&[Bucket]` rather than a `&[&Buckets]` made it harder. Can probably make it work but don't wanna block this PR any longer atm